akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -988,6 +994,75 @@ trait FlowOps[+Out, +Mat] {
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam S the type of the state
+   * @tparam T the type of the output elements
+   * @param create a function that creates the initial state
+   * @param f a function that transforms the upstream element and the state into a pair of next state and output element
+   * @param onComplete a function that transforms the ongoing state into an optional output element
+   */
+  def statefulMap[S, T](create: () => S)(f: (S, Out) => (S, T), onComplete: S => Option[T]): Repr[T] =
+    via(new StatefulMap[S, Out, T](create, f, onComplete))
+
+  /**
+   * Transform each stream element with the help of a resource.
+   *
+   * The resource creation function is invoked once when the stream is materialized and the returned resource is passed to
+   * the mapping function for mapping the first element. The mapping function returns a mapped element to emit
+   * downstream. The returned `T` MUST NOT be `null` as it is illegal as stream element - according to the Reactive Streams specification.
+   *
+   * The `close` function is called only once when the upstream or downstream finished, You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Early completion can be done with combination of the [[takeWhile]] operator.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * You can configure the default dispatcher for this Source by changing the `akka.stream.materializer.blocking-io-dispatcher` or
+   * set it for a given Source by using [[ActorAttributes]].
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam S the type of the resource
+   * @tparam T the type of the output elements
+   * @param create a function that creates the initial state
+   * @param f a function that transforms the upstream element and the state into a pair of next state and output element
+   * @param close a function that transforms the ongoing state into an optional output element
+   */
+  def mapWithResource[S, T](create: () => S)(f: (S, Out) => T, close: S => Option[T]): Repr[T] =
+    via(
+      new StatefulMap[S, Out, T](create, (s, out) => (s, f(s, out)), close)
+        .withAttributes(DefaultAttributes.mapWithResource))
```

Comment:
should the `close` be `S => Unit` ? I think keep it as `S => Option[T]` is more flexible

Reply:
On the fence, option is more flexible but Unit is more obvious for a resource that maps elements and doesn't "create" new elements out of thin air?

Reply:
`S => Unit` would be more consistent with the existing `unfoldResource`

Reply:
otherName `using` `bracket`

Reply:
After chat with some friends at wechat group, they do think `Option` or `Optional` is more flexible, and both `None` and `Optioanl.empty` will cause no allocation.

Reply:
I have updated the PR to return `close: R => Unit` but still think returning `Option` is more flexible, especially when using the `R` as `resource with some state`.

Reply:
Yes but the `Option` one can be more flexiable, `foldResource` 、`foldResourceAsync` should use the `close:Resource => Unit` but not sure about `mapWithResource`.
```scala
    Source(List(ByteString ...))
      .mapWithResource(() => openFile)((file, data) => {
        file.writeData
        None
      }, file => {
        val fileName = file.getName
        file.close()
        Some(fileName)
      })
      .collect {
        case Some(fileName) => fileName
      }
      .to(Sink.headOption)
      .run(...)
...
```
which I can then get the file name